instrumentation/netty/netty-4.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/NettyChannelPipelineInstrumentation.java

```diff
@@ -106,39 +119,78 @@ public static void addHandler(
-        if (handler instanceof HttpServerCodec) {
-          pipeline.addAfter(
-              name, HttpServerTracingHandler.class.getName(), new HttpServerTracingHandler());
-        } else if (handler instanceof HttpRequestDecoder) {
-          pipeline.addAfter(
-              name,
-              HttpServerRequestTracingHandler.class.getName(),
-              new HttpServerRequestTracingHandler());
-        } else if (handler instanceof HttpResponseEncoder) {
-          pipeline.addAfter(
-              name,
-              HttpServerResponseTracingHandler.class.getName(),
-              new HttpServerResponseTracingHandler());
-        } else
+      ChannelHandler ourHandler = null;
+      // Server pipeline handlers
+      if (handler instanceof HttpServerCodec) {
+        ourHandler = new HttpServerTracingHandler();
+      } else if (handler instanceof HttpRequestDecoder) {
+        ourHandler = new HttpServerRequestTracingHandler();
+      } else if (handler instanceof HttpResponseEncoder) {
+        ourHandler = new HttpServerResponseTracingHandler();
         // Client pipeline handlers
-        if (handler instanceof HttpClientCodec) {
-          pipeline.addAfter(
-              name, HttpClientTracingHandler.class.getName(), new HttpClientTracingHandler());
-        } else if (handler instanceof HttpRequestEncoder) {
-          pipeline.addAfter(
-              name,
-              HttpClientRequestTracingHandler.class.getName(),
-              new HttpClientRequestTracingHandler());
-        } else if (handler instanceof HttpResponseDecoder) {
-          pipeline.addAfter(
-              name,
-              HttpClientResponseTracingHandler.class.getName(),
-              new HttpClientResponseTracingHandler());
+      } else if (handler instanceof HttpClientCodec) {
+        ourHandler = new HttpClientTracingHandler();
+      } else if (handler instanceof HttpRequestEncoder) {
+        ourHandler = new HttpClientRequestTracingHandler();
+      } else if (handler instanceof HttpResponseDecoder) {
+        ourHandler = new HttpClientResponseTracingHandler();
+      }
+
+      if (ourHandler != null) {
+        try {
+          pipeline.addAfter(name, ourHandler.getClass().getName(), ourHandler);
+          InstrumentationContext.get(ChannelHandler.class, ChannelHandler.class)
+              .putIfAbsent(handler, ourHandler);
+        } catch (IllegalArgumentException e) {
+          // Prevented adding duplicate handlers.
         }
-      } catch (IllegalArgumentException e) {
-        // Prevented adding duplicate handlers.
+      }
+    }
+  }
+
+  public static class ChannelPipelineRemoveAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void removeHandler(
+        @Advice.This ChannelPipeline pipeline, @Advice.Argument(0) ChannelHandler handler) {
+      ChannelHandler ourHandler =
+          InstrumentationContext.get(ChannelHandler.class, ChannelHandler.class).get(handler);
```

Comment:
Shouldn't we also remove the handler from the `ContextStore`?

Reply:
Thanks, removed it.